lab_5_scrapper/scrapper.py

```diff
@@ -204,13 +353,114 @@ def prepare_environment(base_path: Union[pathlib.Path, str]) -> None:
+
+            with open(pathlib.Path(__file__).parent.parent / 'tmp' / 'recursive_crawler.json', 'r', encoding='utf-8') as infile:
+                data = json.load(infile)
+
+            self.urls = data['urls']
+            self.num_of_urls = data['num_visited_urls']
+            self.possible_urls = data['possible_urls']
+            self.visited_urls = data['visited_urls']
+
+    def save_info(self) -> None:
+        data = {
+            'num_visited_urls': self.num_of_urls,
+            'urls': self.urls,
+            'possible_urls': self.possible_urls,
+            'visited_urls': self.visited_urls
+        }
+        with open(pathlib.Path(__file__).parent.parent / 'tmp' / 'recursive_crawler.json', 'w',
+                  encoding='utf-8') as file:
+            json.dump(data, file, indent=4)
+
+    def find_articles(self) -> None:
+        url = self.possible_urls[self.num_of_urls-1]
+
+        if url not in self.visited_urls:
+            self.visited_urls.append(url)
+
+        response = make_request(url, self.config)
+        article_bs = BeautifulSoup(response.text, 'lxml')
+
+        links = article_bs.find_all('a')
+        articles = []
+        for div in article_bs.find(class_="card-body").find_all("div", {"class": "title"}):
+            for link in div.select("a"):
+                articles.append(link['href'])
+
+        for i in links:
+            if i.get('href'):
+                if 'https' in i.get('href'):
+                    link_url = i.get('href')
+                else:
+                    link_url = 'https://scientificrussia.ru' + i.get('href')
+
+                if link_url not in self.possible_urls:
+                    self.possible_urls.append(link_url)
+                if link_url not in self.urls and i.get('href') in articles:
+                    self.urls.append(link_url)
+
+                if len(self.urls) >= self.config.get_num_articles():
+                    self.save_info()
+                    return
+
+        self.num_of_urls += 1
+        self.find_articles()
 
 
 def main() -> None:
     """
     Entrypoint for scrapper module.
     """
+    conf = Config(CRAWLER_CONFIG_PATH)
+    crawler = Crawler(conf)
+    crawler.find_articles()
+    prepare_environment(ASSETS_PATH)
+
+    for id_num, url in enumerate(crawler.urls, 1):
+        parser = HTMLParser(url, id_num, conf)
+        article = parser.parse()
+        if isinstance(article, Article):
+            to_raw(article)
+            to_meta(article)
+
+
+def recursive_main() -> None:
+    conf = Config(CRAWLER_CONFIG_PATH)
+    crawler = CrawlerRecursive(conf)
+    crawler.find_articles()
+    prepare_environment(ASSETS_PATH)
+
+    for id_num, url in enumerate(crawler.urls, 1):
+        parser = HTMLParser(url, id_num, conf)
+        article = parser.parse()
+        if isinstance(article, Article):
+            to_raw(article)
+            to_meta(article)
 
 
 if __name__ == "__main__":
-    main()
+    recursive_main()
```

Comment:
keep main. it is better for CI. for demonstration in class, you will change what you call.

Reply:
did it